File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-valid-material-opacity.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveJsxElementType } from "../../utils/resolve-jsx-element-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { MAXIMUM_MATERIAL_OPACITY, MINIMUM_MATERIAL_OPACITY } from "./constants.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { hasR3fRuntimeImport } from "./utils/has-r3f-runtime-import.js";
import { isR3fHostIntrinsic } from "./utils/is-r3f-host-intrinsic.js";

export const r3fValidMaterialOpacity = defineRule({
  id: "r3f-valid-material-opacity",
  title: "R3F material opacity outside its normalized range",
  category: "Correctness",
  tags: ["react-jsx-only"],
  severity: "warn",
  recommendation: "Keep material opacity in the normalized [0, 1] range",
  create: (context: RuleContext) => {
    let importsReactThreeFiber = false;
    return {
      Program(node: EsTreeNodeOfType<"Program">) {
        importsReactThreeFiber = hasR3fRuntimeImport(node, context.scopes);
      },
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        const elementType = resolveJsxElementType(node);
        if (
          !importsReactThreeFiber ||
          !isR3fHostIntrinsic(node) ||
          !elementType?.endsWith("Material") ||
          node.attributes.some((attribute) => isNodeOfType(attribute, "JSXSpreadAttribute"))
        ) {
          return;
        }
        const opacityAttribute = getAuthoritativeJsxAttribute(node.attributes, "opacity");
        if (
          !opacityAttribute?.value ||
          !isNodeOfType(opacityAttribute.value, "JSXExpressionContainer") ||
          isNodeOfType(opacityAttribute.value.expression, "JSXEmptyExpression")
        ) {
          return;
        }
        const opacity = getStaticNumber(opacityAttribute.value.expression, context.scopes);
        if (
          opacity === null ||
          (opacity >= MINIMUM_MATERIAL_OPACITY && opacity <= MAXIMUM_MATERIAL_OPACITY)
        ) {
          return;
        }
        context.report({
          node: opacityAttribute,
          message: `Material opacity is ${String(opacity)}, but Three.js opacity uses the normalized [0, 1] range`,
        });
      },
    };
  },
});

File: native/oxlint/rules/r3f-valid-material-opacity.rs
use oxc_ast::{ast::JSXAttributeValue, AstKind};
use oxc_diagnostics::OxcDiagnostic;

use crate::{
    context::{ContextHost, LintContext},
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const MINIMUM_MATERIAL_OPACITY: f64 = 0.0;
const MAXIMUM_MATERIAL_OPACITY: f64 = 1.0;

#[derive(Debug, Default, Clone)]
pub struct R3FValidMaterialOpacity;

impl RuleMeta for R3FValidMaterialOpacity {
    const NAME: &'static str = "r3f-valid-material-opacity";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Correctness;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Validate React Three Fiber material opacity.",
    };
}

impl Rule for R3FValidMaterialOpacity {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        if file_is_non_react_jsx_dialect(ctx) || !has_r3f_runtime_import(ctx) {
            return;
        }
        for node in ctx.nodes().iter() {
            let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
                continue;
            };
            let Some((element_type, _)) = resolve_jsx_element_type(opening_element, ctx) else {
                continue;
            };
            if !element_type.ends_with("Material")
                || !element_type.chars().next().is_some_and(char::is_lowercase)
                || element_type.contains('-')
                || opening_element.attributes.iter().any(|attribute| {
                    matches!(
                        attribute,
                        oxc_ast::ast::JSXAttributeItem::SpreadAttribute(_)
                    )
                })
            {
                continue;
            }
            let Some(attribute) = get_authoritative_jsx_attribute(opening_element, "opacity", true)
            else {
                continue;
            };
            let Some(JSXAttributeValue::ExpressionContainer(container)) = &attribute.value else {
                continue;
            };
            let Some(expression) = container.expression.as_expression() else {
                continue;
            };
            let Some(opacity) = resolve_static_number(expression, ctx) else {
                continue;
            };
            if (MINIMUM_MATERIAL_OPACITY..=MAXIMUM_MATERIAL_OPACITY).contains(&opacity) {
                continue;
            }
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "Material opacity is {opacity}, but Three.js opacity uses the normalized [0, 1] range"
                ))
                .with_label(attribute.span),
            );
        }
    }
}
